R/process_sc_all_care_home.R

```diff
@@ -237,72 +233,98 @@ process_sc_all_care_home <- function(
 
   # Continuous Care Home Stays
   # Stay will be continuous as long as the admission date is the next day or
-  # earlier than the previous discharge date
-
-  ch_markers <- matched_deaths_data %>%
-    # ch_chi_cis
+  # earlier than the previous discharge date.
+  # creates a CIS  flag for CHI across all of scotland
+  # and a CIS for social care ID and sending location for just that LA
+  ch_chi_markers <- matched_deaths_data %>%
+    # uses the chi to flag continuous stays. Will flag cases even if in another LA
     dplyr::group_by(.data[["chi"]]) %>%
-    dplyr::mutate(
-      continuous_stay_chi = tidyr::replace_na(
-        .data[["ch_admission_date"]] <= dplyr::lag(
-          .data[["ch_discharge_date"]]
-        ) + lubridate::days(1L),
-        TRUE
-      ),
-      ch_chi_cis = cumsum(.data[["continuous_stay_chi"]])
+    # create variable for previous discharge date + 1 day
+    dplyr::mutate(previous_discharge_date_chi = dplyr::lag(.data[["ch_discharge_date"]]) + lubridate::days(1L)) %>%
+    # TRUE/FALSE flag for if admission date is before or equal to previous discharge date + 1 day
+    dplyr::mutate(continuous_stay_flag_chi = tidyr::replace_na(.data[["ch_admission_date"]] <= previous_discharge_date_chi, FALSE)) %>%
+    # different to code in above sections.
+    # we want to uniquely identify all cases where the flag is FALSE. and only the first case where the flag is TRUE
+    # to do this create a variable of the flag in the previous row
+    dplyr::mutate(previous_continuous_stay_flag_chi = tidyr::replace_na(dplyr::lag(.data[["continuous_stay_flag_chi"]]), FALSE)) %>%
+    dplyr::mutate(continuous_stay_chi = ifelse(continuous_stay_flag_chi == FALSE |
+      (continuous_stay_flag_chi == TRUE & previous_continuous_stay_flag_chi == FALSE), FALSE, TRUE)) %>%
+    dplyr::group_by(
+      .data[["chi"]],
+      .data[["continuous_stay_chi"]]
     ) %>%
-    dplyr::ungroup() %>%
-    # ch_sc_id_cis
-    # uses the social care id and sending location so can be used for
-    # episodes that are not attached to a CHI number
-    # This will restrict continuous stays to each Local Authority
-    dplyr::group_by(.data[["social_care_id"]], .data[["sending_location"]]) %>%
-    dplyr::mutate(
-      continuous_stay_sc = tidyr::replace_na(
-        .data[["ch_admission_date"]] <= dplyr::lag(
-          .data[["ch_discharge_date"]]
-        ) + lubridate::days(1L),
-        TRUE
-      ),
-      ch_sc_id_cis = cumsum(.data[["continuous_stay_sc"]])
+    # gives cases their unique CIS identifier
+    dplyr::mutate(ch_chi_cis = ifelse(continuous_stay_chi == FALSE, dplyr::row_number(), NA)) %>%
+    dplyr::group_by(
+      .data[["social_care_id"]],
+      .data[["sending_location"]]
     ) %>%
-    dplyr::ungroup()
+    # fills in CIS identifier for all cases
+    tidyr::fill(ch_chi_cis, .direction = c("down"))
```

Comment:
I think this is working more like how we want it to. However, due to using the lag function and NA being assigned to the top row, it means that for some cases the first episode is being assigned 1 and the next episode would be assigned 2 even if it is part of the same stay. Its due to the missing date on the row before that it is not calculating this properly. The cases after the first row of each episode is working as we would expect.

I have shown @SwiftySalmon a few examples on a call and she is going to have another look at the logic. Other than this part of care homes, i think everything else is working as we would expect. Well done Megan for all your hard work on this complex PR